artemis-server/src/main/java/org/apache/activemq/artemis/core/persistence/impl/journal/AbstractJournalStorageManager.java

```diff
@@ -1488,7 +1494,13 @@ public synchronized void start() throws Exception {
 
       beforeStart();
 
-      singleThreadExecutor = executorFactory.getExecutor();
+      ThreadFactory tFactory = AccessController.doPrivileged(new PrivilegedAction<ThreadFactory>() {
+         @Override
+         public ThreadFactory run() {
+            return new ActiveMQThreadFactory("ActiveMQ-journal-server-" + this.toString(), true, ClientSessionFactoryImpl.class.getClassLoader());
+         }
+      });
+      singleThreadExecutor = Executors.newSingleThreadExecutor(tFactory);
```

Comment:
nope.. that's wrong...

executorFactory.getExecutor() is returning on thread executor from the pool.


it won't always be the same thread.. but it will always be the same context.. this patch is not valid.


in what situation do you see a deadlock.


hornetq it might be different.. I would need a test to be able to accept a patch here. we should reuse the thread from the pool always.

Reply:
Can you please close your PR? The executor factory is providing the same thing you suggested here. it won't fix anything. .just will create more threads and problems.

Reply:
What if the pool is full?
In our case, the pool is a 60 thread fixed pool.
One of the thread is doing page cleanup, and try to exit paging state, it holds the lock in paging store. All other 59 threads is blocked on the lock, trying to page.
While cleanup, we need to store bookmark in journal for each page subscription, then wait until completed.
In log, stored equals to storeLineUp, but there are pending tasks(there are going to count down latch cleanup thread is waiting on), the deadlock happened.
```
16:44:28,930 AMQ222024: Could not complete operations on IO context OperationContextImpl [1251391301] [minimalStore=1, storeLineUp=2, stored=2, minimalReplicated=0, replicationLineUp=0, replicated=0, paged=0, minimalPage=0, pageLineUp=0, errorCode=-1, errorMessage=null, executorsPending=3, executor=OrderedExecutor(tasks=[org.apache.activemq.artemis.core.persistence.impl.journal.OperationContextImpl$1@4d09259, org.apache.activemq.artemis.core.persistence.impl.journal.OperationContextImpl$1@54b73dc4, org.apache.activemq.artemis.core.persistence.impl.journal.OperationContextImpl$1@640495d4])]
```

Reply:
We believe this happens when
- using fixed thread pool which is the default, and
- an address has many producers, more then thread pool's size, and
- the address is about to exit paging state

Reply:
Can I see a thread dump ? Maybe there is a better fix. 

A test would be best.  But a thread dump would be ok as long as you tell me the version (or got commit) it is associated.

Reply:
We are running 2.4.0
See https://issues.apache.org/jira/browse/ARTEMIS-1700
There is a artemis.log attached.

Reply:
[artemis.log](https://github.com/apache/activemq-artemis/files/1753723/artemis.log)
Please see the thread dump file in attachment.

Reply:
https://github.com/apache/activemq-artemis/blob/d6d895c558cc104475188d942473771418b5e3e6/artemis-server/src/main/java/org/apache/activemq/artemis/core/server/group/impl/LocalGroupingHandler.java#L109-L117

@clebertsuconic the comment said we need an executor out side the pool

Reply:
@shoukunhuai there's another pool to be used with IO.. iothreadpool.. this is using the wrong pool.. besides we shouldn't wait forever to stop... I'm sending another PR.

Reply:
if you could please comment on #1899